src/main/java/bsh/Operators.java

```diff
@@ -399,9 +401,62 @@ static Object bigIntegerBinaryOperation(BigInteger lhs, BigInteger rhs, int kind
                 "Unimplemented binary integer operator");
     }
 
+    // returns Object covering both Float return types
+    static Object floatBinaryOperation(float lhs, float rhs, int kind)
+            throws UtilEvalError
+    {
+        switch(kind)
+        {
+            // arithmetic
+            case PLUS:
+                if ( lhs > 0d && (Float.MAX_VALUE - lhs) < rhs )
+                    break;
+                return lhs + rhs;
+
+            case MINUS:
+                if ( lhs < 0d && (-Float.MAX_VALUE - lhs) > -rhs )
+                    break;
+                return lhs - rhs;
+
+            case STAR:
+                if ( lhs != 0 && Float.MAX_VALUE / lhs < rhs )
+                    break;
+                return lhs * rhs;
+
+            case SLASH:
+                return lhs / rhs;
+
+            case MOD:
+            case MODX:
+                return lhs % rhs;
+
+            case POWER:
+            case POWERX:
+                double check = Math.pow(lhs, rhs);
+                if ( Double.isInfinite(check) )
+                    break;
+                return check;
+
+            // can't shift floating-point values
+            case LSHIFT:
+            case LSHIFTX:
+            case RSIGNEDSHIFT:
+            case RSIGNEDSHIFTX:
+            case RUNSIGNEDSHIFT:
+            case RUNSIGNEDSHIFTX:
+                throw new UtilEvalError("Can't shift float values");
+
+        }
+        if ( OVERFLOW_OPS.contains(kind) )
+            return bigDecimalBinaryOperation(BigDecimal.valueOf(lhs), BigDecimal.valueOf(rhs), kind);
```

Comment:
I wonder if we shouldn't first overflow to double here

Reply:
Can we just stick with Java math please?  Some of us depend on it.